### PR TITLE
internal/xdsclient: xdsclient transport logs should indicate the authority

### PR DESCRIPTION
**Pull request by my4-dev**
For more information, please refer to #6091 .

RELEASE NOTES: none

### CODE REVIEW

File: xds/internal/xdsclient/logging.go
Comment (arvindbr8) [inline xds/internal/xdsclient/logging.go:1]: If you do by the approach I suggested earlier none of this needs to change

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
Hey @my4-dev. Thanks for the PR. Seems like someone is raising the bar on turn around time.  But I have an overall comment on the diff.

I would take a different approach which is more idiosyncratic with the usage of `prefixLogger` in other places in the repo. I feel like we can go about without having to expose a new API (AddPrefix) to the `preflixLogger`. 

How about we simple create a new instance of `prefixLogger` [here](https://github.com/grpc/grpc-go/blob/8628e075df225d916eaeaec1488af4dcc02805ec/xds/internal/xdsclient/clientimpl_authority.go#L106-L112) like so 

```
        // Make a new authority since there's no existing authority for this config.
+       loggerPrefix := fmt.Sprintf("%s [%%s]", prefix)
        ret, err := newAuthority(authorityArgs{
                serverCfg:          config,
                bootstrapCfg:       c.config,
                serializer:         c.serializer,
                resourceTypeGetter: c.resourceTypes.get,
                watchExpiryTimeout: c.watchExpiryTimeout,
-               logger:             c.logger,
+               logger:             internalgrpclog.NewPrefixLogger(logger, fmt.Sprintf(loggerPrefix, c, config.ServerURI)),
        })

```

This way we don't require any additional changes anywhere else in the code. Let me know what you think.

**Comment by arvindbr8** [comment]
Optional: While you are there, it would be great if you could you also help cleanup these incorrect usages of `logger`

- [here](https://github.com/grpc/grpc-go/blob/master/xds/internal/xdsclient/clientimpl_watchers.go#L159)
- [here](https://github.com/grpc/grpc-go/blob/master/xds/internal/xdsclient/clientimpl_watchers.go#L164)
- [here](https://github.com/grpc/grpc-go/blob/master/xds/internal/xdsclient/clientimpl_watchers.go#L183)


The write way would be by calling the xdsclient's logger -- which means you could get away with a simple 

```go
- logger.Warningf("Watch registered for name %q of type %q which is already registered", rType.TypeName(), resourceName) 
+ c.logger.Warningf("Watch registered for name %q of type %q which is already registered", rType.TypeName(), resourceName)

```

**Comment by my4-dev** [comment]
Hi, @arvindbr8 ! Thanks for your advice!
I have some opinions to the approach you suggested earlier.

1. `internalgrpclog.NewPrefixLogger`'s first argument type is `DepthLoggerV2`, so we couldn't use `NewPrefixLogger` like you suggested without changing this API.
2. I have added the usage of `prefixLoggerAuthorityArgs` at `newAuthority` method in authoriy.go because we want to confirm the information of management servers whenever we initiate an authority.

File: xds/internal/xdsclient/logging.go [hunk 1]
```diff
@@ -25,10 +25,30 @@ import (
...
-	return internalgrpclog.NewPrefixLogger(logger, fmt.Sprintf(prefix, p))
+func prefixLoggerClient(p *clientImpl) *internalgrpclog.PrefixLogger {
+	if p.logger != nil {
+		return p.logger.AddPrefix(fmt.Sprintf(prefixC, p))
+	}
+	return internalgrpclog.NewPrefixLogger(logger, fmt.Sprintf(prefixC, p))
+}
+
+func prefixLoggerAuthorityArgs(a *authorityArgs) *internalgrpclog.PrefixLogger {
+	if a.serverCfg == nil || a.serverCfg.ServerURI == "" {
```
[the remark this answers is not in the document]
Comment (my4-dev) [reply xds/internal/xdsclient/logging.go:43]: I implemented this logic in order not to occur a system error like a null pointer exception.
The logic where `serverCfg` is defined  is so complicated that I couldn't confirm the possibility that `serverCfg` becomes nil.
Do you think this is unneccessary?

**Comment by arvindbr8** [comment]
SG. Let me know when you want me to take another look at it.

**Comment by arvindbr8** [comment]
Assigning it back to you @my4-dev. Seems like this might still be in progress. Let me know if you still have questions.

**Comment by my4-dev** [comment]
@arvindbr8 : 
Thank you for assigning back. I don't have any questions.
If you or other member don't have any other suggestions or opinions after reviewing this PR, it would be ready to be merged.

**Comment by arvindbr8** [comment]
sure. taking another look

File: xds/internal/xdsclient/logging.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (arvindbr8) [reply xds/internal/xdsclient/logging.go:43]: Yeah, i dont think there would be case when serverCfg is nil or the ServerURI is nil when the authority is already instantiated.

**Review by arvindbr8 — CHANGES_REQUESTED** [review]
Hi, @arvindbr8 ! Thanks for your advice!
I have some opinions to the approach you suggested earlier.

> 1. internalgrpclog.NewPrefixLogger's first argument type is DepthLoggerV2, so we couldn't use NewPrefixLogger like you suggested without changing this API.

The diff i suggested in my previous comment seems to checkout in my local. Is that a type mismatch error that you see?

> 2. I have added the usage of prefixLoggerAuthorityArgs at newAuthority method in authoriy.go because we want to confirm the information of management servers whenever we initiate an authority.

All the information required about the mgmt server should be available to you from the bootstrap config. 

let me know this doesnt make sense to you. We could discuss more about it in the related issue.

**Comment by my4-dev** [comment]
Hi, @arvindbr8 !

> The diff i suggested in my previous comment seems to checkout in my local. Is that a type mismatch error that you see?

I'm sorry, I tried again and could confirm that a type mismatch error didn't occur.
I might have tried with `c.logger` instead of `logger`🙏
I think that it is unnecessary to expose a new API (AddPrefix) to the `preflixLogger` too.

> All the information required about the mgmt server should be available to you from the bootstrap config.

If we forget setting logger appropriately when we initiate `authority` by using `newAuthority` in other situation in the future, mgmt server information don't be outputted by logger.
Do we need to consider this situation?

**Comment by arvindbr8** [comment]
Sorry for the delay. I was out of office. Taking a look now.

**Comment by arvindbr8** [comment]
I dont think adding another prefix to our prefix logger would be the right way to go. Let me clone your repo and check the type error from my side.

**Comment by arvindbr8** [comment]
Hi @my4-dev, Could you check this diff out. My idea to implement this would be along this diff. 

On a high level, the logger that is passed to authority.go and transport.go needs to have the new prefix as per #6091. I dont believe adding a  `AddPrefix` method is the right way to go about it. Please take a look and let me know if this PR can align more on those lines. 

If you think its too much work, feel free to discard this PR and one of us could definitely take this up.

**Comment by my4-dev** [comment]
@arvindbr8 :
I'm sorry, I want one of your team members to take this up🙇‍♂️

**Pull request closed without merging**